### PR TITLE
style: fix the 156 ruff errors that keep pre-commit.ci red on main (#244)

### LINKED ISSUE #244: pre-commit.ci has been red on main since at least 2026-07-13 (156 ruff errors, none auto-fixable)
## Description

`pre-commit.ci` is failing on `main`, and because of that **every open pull request is red** on the `pre-commit.ci - pr` check regardless of what it changes.

The failure is the `ruff` hook. It is not caused by any individual PR — the offending code is all pre-existing on `main`.

### Evidence that `main` itself is red

`GET /repos/openclimatefix/graph_weather/commits/main/status` returns `"state": "failure"` with a single status, `pre-commit.ci - push`. The last several commits on `main` are all in the same state:

| commit | date | combined status |
| --- | --- | --- |
| `a23d8c5` | 2026-07-27 | failure |
| `905002f` | 2026-07-26 | failure |
| `1adae8b` | 2026-07-20 | failure |
| `36e4934` | 2026-07-13 | failure |
| `95097fd` | 2026-07-13 | failure |

So this has been broken since at least 2026-07-13.

### What ruff actually reports

Running the exact version pinned in `.pre-commit-config.yaml` (`astral-sh/ruff-pre-commit` rev `v0.15.15`) against `main` at `a23d8c5`:

```
$ ruff check --output-format=concise .
Found 156 errors.
```

By rule:

| count | rule | name |
| --- | --- | --- |
| 33 | D102 | undocumented-public-method |
| 33 | D205 | missing-blank-line-after-summary |
| 29 | D107 | undocumented-public-init |
| 27 | D101 | undocumented-public-class |
| 12 | E501 | line-too-long |
| 10 | D100 | undocumented-public-module |
| 5 | D417 | undocumented-param |
| 4 | D103 | undocumented-public-function |
| 3 | D104 | undocumented-public-package |

Spread over 33 files, concentrated in:

```
29  graph_weather/models/fengwu_ghr/layers.py
17  graph_weather/models/aurora/model.py
10  graph_weather/models/cafa/factorize.py
 9  graph_weather/models/weathermesh/weathermesh2.py
 8  train/lora.py
 7  graph_weather/models/cafa/encoder.py
 6  graph_weather/models/weathermesh/{processor,encoder,decoder}.py
 6  graph_weather/models/losses.py
 ...
```

### Why `autofix_prs` does not clear it

The hook already runs with `args: [--fix]`, so anything auto-fixable is fixed before the failure is reported. `ruff check --fix --diff` on `main` produces an empty diff — **0 of the 156 are auto-fixable**. They all need docstrings written by hand (plus 12 line wraps).

### A note for anyone reading the CI log

The `results.pre-commit.ci` page truncates long output. All three runs I looked at (`main`, and two PRs) cut off at the same line with `(output truncated)`, showing only the first 54 errors. The real count is 156 — worth knowing before sizing the fix.

## Fix

I have a branch that takes this to `All checks passed!` — docstrings and line wraps only, with every changed file verified AST-identical to `main` once string-literal expression statements are discarded. I opened it as #245 and then closed it again: 1,200 lines across 33 files is too large a change to push at you unasked.

It sits on my fork at `yasumorishima:fix/ruff-docstrings`. Say the word and I will reopen it, split it per-module, or leave it alone — whichever suits you.

Writing those docstrings also surfaced thirteen existing docstrings that describe behavior the code does not have, and seven genuine bugs. Both lists are in #245 if they are useful to you.

### PR DESCRIPTION
**Pull request by yasumorishima**
# Pull Request

## Description

`pre-commit.ci` has been failing on `main` since at least 2026-07-13, which makes the `pre-commit.ci - pr` check red on **every** open pull request regardless of what it changes. The `ruff` hook reports 156 errors and `ruff check --fix --diff` produces an empty diff, so `autofix_prs` cannot clear any of them.

This adds the missing Google-style docstrings, inserts the blank line after the summary line where `D205` asked for it, fills in the two undocumented `efficient_batching` arguments, and wraps the 12 over-long lines.

```
before:  Found 156 errors.
after:   All checks passed!
```

Fixes #244

## No behavior changes

Every changed file was compared against `upstream/main` by parsing both versions with `ast`, discarding **all** string-literal expression statements (docstrings, and one module-level notes block that is written as a bare string), and diffing `ast.dump` of the result:

- executable AST identical: **33/33**
- comment text identical after whitespace normalization: **33/33**
- `black 26.5.1 --line-length 100 --check .` → 115 files would be left unchanged
- no non-Python file is touched

## Docstrings that were wrong, not just missing

Thirteen existing statements did not match the code. They are corrected rather than reformatted, each verified against the source:

| file | said | actually |
| --- | --- | --- |
| `cafa/processor.py` | `forward` takes `(batch, height, width, channels)` | first op is `rearrange(x, "b c h w -> b h w c")`, so it takes `(batch, channels, height, width)` |
| `cafa/encoder.py` | argument `input_channel` | the parameter is `input_channels` — this is what `D417` was flagging |
| `cafa/decoder.py` | `FactorizedTransformer` | the package defines `CaFAProcessor` and `FactorizedTransformerBlock` |
| `train/era5.py` | `LitFengWuGHR.__init__` takes `aux_dim`, `hidden_dim`, `num_blocks` | none of the three exist; six real parameters were undocumented |
| `train/era5.py` | `self.model` is a `GraphWeatherForecaster` | it is a `MetaModel` |
| `aurora/processor.py` | `attention_dropout`, `layer_norm_eps`, `num_cross_attention_layers`, `max_seq_len`, `qk_head_dim` take effect | `PerceiverProcessor.__init__` reads only `input_dim`, `d_model`, `num_attention_heads`, `hidden_dropout`, `activation_fn`, `num_self_attention_layers`, `latent_dim`; the rest are validated in `__post_init__` at most |
| `aurora/processor.py` | `input_dim` matches the Swin3D output | `Swin3DEncoder` defaults to `96`, `ProcessorConfig` to `256` |
| `aurora/model.py` | `spatial_correlation_loss` averages over neighboring pairs | `torch.mean(nearby_mask * ...)` divides by every pair, masked-out zeros included |
| `aurora/model.py` | `PointEncoder` is order-invariant | point-wise MLPs plus addition are permutation-**equivariant** |
| `aurora/model.py` | `forward` returns `num_points` rows | the encoder truncates to `max_seq_len`, so it is `min(num_points, max_seq_len)` |
| `aurora/decoder.py` | the transposed convolution upscales | `kernel_size=3, stride=1, padding=1` preserves D, H and W; only the channel count changes |
| `fengwu_ghr/layers.py` | `LoRAModule` replaces linear layers | `setattr(model, name, ...)` with the dotted names from `named_modules` only substitutes direct attributes |
| `fengwu_ghr/layers.py` | the batcher cuts the image into sub-images | `"b c (h s_h) (w s_w) -> (b s_h s_w) c h w"` puts `s_h` on the fast axis, so each sub-image is a strided sample, not a contiguous tile |

Two more were adjusted for accuracy: `posemb_sincos_2d` documented `dim` as any multiple of 4, but `dim == 4` passes the assertion and makes `dim // 4 - 1` zero; and the `weathermesh2` implementation note gave the NATTEN window as `depth, width, height` while the encoder hands the processor a `B D H W C` latent, so it is `(depth, height, width)` as every other docstring in that package says (both are 7 by default, so nothing observable changes).

## Bugs found while writing this, left alone

Documenting these modules surfaced seven real defects. None are touched here, because this change is docstrings only — but they are worth knowing about, and I am happy to open follow-up PRs for any of them if that would help:

1. `weathermesh/{encoder,decoder,processor,weathermesh2}.py` — all four `to_json` methods call `dacite.asdict`, which `dacite` does not export (`from_dict` is the only converter). They raise `AttributeError`. `dataclasses.asdict` is the drop-in.
2. `fengwu_ghr/layers.py` — `LoRAModule` iterates `named_modules()` and calls `setattr(model, name, ...)` with dotted names, so nested linear layers keep running unchanged; it also mutates `_modules` during that traversal.
3. `fengwu_ghr/layers.py` — `LoRALayer.forward` computes `self.B @ self.A @ x`, which does not apply to the trailing `in_features` axis for inputs with leading dimensions.
4. `aurora/model.py` — `spatial_correlation_loss` does `cdist` on `(B*N, 2)` and then `view(B, N, N)`; the element counts only match when `B == 1`.
5. `aurora/processor.py` — the 4D branch of `PerceiverProcessor.forward` uses `rearrange(x, "b s h w -> b (s h w) c")`, where `c` is not bound.
6. `layers/constraint_layer.py` — the constructor defaults `constraint_type="none"`, which `forward` has no branch for, so the default raises `ValueError`.
7. `weathermesh2.py` — `self.processors` is a plain list in both constructor branches, so those modules are invisible to `state_dict()`, `.to(device)` and `.train()`.

## How Has This Been Tested?

- [x] Yes

- `ruff check .` with the pinned `v0.15.15` → `All checks passed!` (was `Found 156 errors.`)
- `black 26.5.1 --line-length 100 --check .` → unchanged
- the AST and comment-text comparison described above, over all 33 changed files
- `pytest` is red on this branch, but it is red on `main` too: it is the collection-time segfault of #232 (exit 139 before any test runs), which #241 addresses. Docstrings cannot cause it, and the executable AST here is identical to `main`.

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works — the `ruff` hook is the test; there is no runtime behavior to cover
- [x] I have checked my code and corrected any misspellings

Happy to split this per-module if a single sweep is awkward to review — just say the word.

### CODE REVIEW

**Comment by yasumorishima** [comment]
Closing this. On reflection, dropping 1,200 lines of docstrings across 33 files into someone else's repository without being asked for it is not a reasonable thing to do, however green the check turns.

Leaving the findings here in case any of them are useful to you:

- `ruff check .` with your pinned `v0.15.15` reports **156** errors on `main`, not the 54 the `results.pre-commit.ci` page shows — that page truncates and the three runs I looked at all cut off at the same line. None are auto-fixable, so `autofix_prs` cannot clear them.
- Thirteen existing docstrings state things the code does not do. The clearest: `cafa/processor.py` documents `forward` as taking `(batch, height, width, channels)` while the first operation is `rearrange(x, "b c h w -> b h w c")`; `train/era5.py` documents `aux_dim`, `hidden_dim` and `num_blocks` on `LitFengWuGHR.__init__`, none of which exist; `aurora/processor.py` describes `attention_dropout`, `layer_norm_eps`, `num_cross_attention_layers` and `qk_head_dim` as taking effect, but `PerceiverProcessor.__init__` never reads them.
- Seven genuine bugs, listed with detail in the description above. The two I would look at first: all four `to_json` methods in `weathermesh/` call `dacite.asdict`, which `dacite` does not export, so they raise `AttributeError`; and `WeatherMesh.processors` is a plain list in both constructor branches, so those modules are invisible to `state_dict()`, `.to(device)` and `.train()`.

The branch stays on my fork at `yasumorishima:fix/ruff-docstrings` if you ever want it, and I am glad to open any of the bug fixes as a small PR if you would find that useful. Sorry for the noise.
